File: go/internal/vectest/cross.go
package vectest

import (
	"testing"

	"github.com/elkpi/oxa/go/ir"
)

// RunCross executes every nonstream cross-protocol vector whose source and
// target endpoints match the two converters' Face() protocol names. Each
// vector composes source decode -> IR -> target encode; the target wire
// output compares structurally and the reported losses are the decode
// losses followed by the encode losses, compared as an unordered set. A
// matched pair with no vectors fails the test.
func RunCross(t *testing.T, source, target Converter) {
	t.Helper()
	root := FindRepoRoot(".")
	if root == "" {
		t.Skip("repo root not found; vector tests skipped (dependency mode)")
	}
	vectors, err := LoadVectors(root, "cross", "nonstream")
	if err != nil {
		t.Fatalf("load cross vectors: %v", err)
	}
	matched := crossVectorsFor(source, target, vectors)
	if len(matched) == 0 {
		t.Fatalf("no cross vectors found for pair %s -> %s", source.Face(), target.Face())
	}
	for _, v := range matched {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			runCrossVector(t, source, target, v)
		})
	}
	t.Logf("vectest: executed %d cross vectors for %s -> %s", len(matched), source.Face(), target.Face())
}

// crossVectorsFor selects the vectors whose endpoints match the pair.
func crossVectorsFor(source, target Converter, vectors []Vector) []Vector {
	var out []Vector
	for _, v := range vectors {
		if v.Source.Protocol == source.Face() && v.Target.Protocol == target.Face() {
			out = append(out, v)
		}
	}
	return out
}

func runCrossVector(t *testing.T, source, target Converter, v Vector) {
	t.Helper()
	if v.Conversion != "protocol-to-protocol" {
		t.Fatalf("cross vector %s has conversion %q, want protocol-to-protocol", v.Name, v.Conversion)
	}
	var out []byte
	var decodeLosses, encodeLosses []ir.Loss
	var err error
	if v.isRequest() {
		var req *ir.Request
		req, decodeLosses, err = source.DecodeRequestWire(v.Input)
		if err == nil {
			out, encodeLosses, err = target.EncodeRequestIR(req)
		}
	} else {
		var resp *ir.Response
		resp, decodeLosses, err = source.DecodeResponseWire(v.Input)
		if err == nil {
			out, encodeLosses, err = target.EncodeResponseIR(resp)
		}
	}
	if err != nil {
		t.Fatalf("cross conversion failed: %v", err)
	}
	if err := CompareJSON(v.ExpectedOut, out); err != nil {
		t.Errorf("expected_output mismatch: %v\nexpected: %s\nactual:   %s", err, v.ExpectedOut, out)
	}
	losses := append(append([]ir.Loss{}, decodeLosses...), encodeLosses...)
	compareLosses(t, v, losses)
}

File: rust/crates/oxa-vectest/src/cross.rs
//! The cross-protocol vector runner (spec/02 §6): each vector composes
//! source decode → IR → target encode; losses concatenate in that order and
//! compare as an unordered set.

use std::path::Path;

use crate::compare::compare_json;
use crate::load::{Vector, find_repo_root, load_vectors};
use crate::run::{Converter, Outcome, Report, compare_reported_losses};

/// Executes every nonstream cross-protocol vector whose source and target
/// endpoints match the two converters' `face()` protocol names, locating the
/// repo root from the current working directory. A matched pair with no
/// vectors fails.
pub fn run_cross(source: &dyn Converter, target: &dyn Converter) -> Result<Outcome, String> {
    run_cross_in(Path::new("."), source, target)
}

/// [`run_cross`] with an explicit starting directory for the repo-root walk.
pub fn run_cross_in(
    dir: &Path,
    source: &dyn Converter,
    target: &dyn Converter,
) -> Result<Outcome, String> {
    let Some(root) = find_repo_root(dir) else {
        return Ok(Outcome::Skipped);
    };
    let vectors = load_vectors(&root, "cross", "nonstream")?;
    let matched = cross_vectors_for(source, target, &vectors);
    if matched.is_empty() {
        return Err(format!(
            "no cross vectors found for pair {} -> {}",
            source.face(),
            target.face()
        ));
    }
    let executed = matched.len();
    let mut report = Report::default();
    for vector in &matched {
        run_cross_vector(source, target, vector, &mut report);
    }
    report.executed = executed;
    Ok(Outcome::Ran(report))
}

/// Selects the vectors whose endpoints match the pair.
pub fn cross_vectors_for<'a>(
    source: &dyn Converter,
    target: &dyn Converter,
    vectors: &'a [Vector],
) -> Vec<&'a Vector> {
    vectors
        .iter()
        .filter(|vector| {
            vector.source.protocol == source.face() && vector.target.protocol == target.face()
        })
        .collect()
}

fn run_cross_vector(
    source: &dyn Converter,
    target: &dyn Converter,
    vector: &Vector,
    report: &mut Report,
) {
    if vector.conversion != "protocol-to-protocol" {
        report.failures.push((
            vector.name.clone(),
            format!(
                "cross vector has conversion {:?}, want protocol-to-protocol",
                vector.conversion
            ),
        ));
        return;
    }
    let Some(expected) = vector.expected_output.as_ref() else {
        report.failures.push((
            vector.name.clone(),
            "cross vector is missing expected_output".to_string(),
        ));
        return;
    };
    let outcome: Result<(serde_json::Value, Vec<oxa_ir::Loss>), String> = (|| {
        let input = vector.input_text();
        let (out, losses) = if vector.is_request() {
            let (req, decode_losses) = source.decode_request_wire(&input)?;
            let (out, encode_losses) = target.encode_request_ir(&req)?;
            (out, [decode_losses, encode_losses].concat())
        } else {
            let (resp, decode_losses) = source.decode_response_wire(&input)?;
            let (out, encode_losses) = target.encode_response_ir(&resp)?;
            (out, [decode_losses, encode_losses].concat())
        };
        let out: serde_json::Value = serde_json::from_str(&out).map_err(|err| err.to_string())?;
        Ok((out, losses))
    })();
    match outcome {
        Ok((out, losses)) => {
            if let Err(err) = compare_json(expected, &out) {
                report.failures.push((
                    vector.name.clone(),
                    format!("expected_output mismatch: {err}"),
                ));
            }
            compare_reported_losses(vector, &losses, report);
        }
        Err(err) => report.failures.push((
            vector.name.clone(),
            format!("cross conversion failed: {err}"),
        )),
    }
}
